energy-integration/mex-governance/src/lib.rs

```diff
@@ -0,0 +1,120 @@
+#![no_std]
+
+multiversx_sc::imports!();
+
+pub mod config;
+pub mod errors;
+pub mod events;
+pub mod views;
+
+use config::FarmEmission;
+use errors::*;
+
+#[multiversx_sc::contract]
+pub trait MEXGovernance:
+    config::ConfigModule
+    + events::EventsModule
+    + energy_query::EnergyQueryModule
+    + week_timekeeping::WeekTimekeepingModule
+    + views::ViewsModule
+{
+    #[init]
+    fn init(
+        &self,
+        reference_emission_rate: BigUint,
+        incentive_token: TokenIdentifier,
+        energy_factory_address: ManagedAddress,
+    ) {
+        self.set_reference_emission_rate(reference_emission_rate);
+        self.set_incentive_token(incentive_token);
+        self.energy_factory_address().set(&energy_factory_address);
+    }
+
+    #[upgrade]
+    fn upgrade(&self) {}
+
+    #[endpoint(vote)]
+    fn vote(&self, votes: MultiValueEncoded<MultiValue2<AddressId, BigUint>>) {
```

Comment:
Again, don't complicate everything with IDs. I only meant use the IDs inside contract storage for optimization, not everywhere.

Reply:
Hmmm, the microservice will probably have this list saved so why not making them compose the TX using the id? Anyway, we could keep everything that relates to the user interaction to still use the addresses. But we introduce a few more extra storage reads.

Reply:
In all my years, they've never asked me for IDs instead of addresses, so I wouldn't count on it.